File: mvn-advent-of-code/src/main/java/advent/twenty_fifteen/Day10ElvesLookElvesSay.java
package advent.twenty_fifteen;

import advent.common.DailyProblem;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class Day10ElvesLookElvesSay implements DailyProblem<Integer, Integer> {
    private final int part1Answer;
    private final int part2Answer;

    public Day10ElvesLookElvesSay(String input) {
        LinkedList<Integer> working = new LinkedList<>();
        for (Character c : input.toCharArray()) {
            working.add(Integer.parseInt(""+c));
        }

        List<Integer> part1 = null;
        for (int i = 0; i != 50; i++) {
            working = lookSay(working);

            if (i == 39) {
                part1 = new LinkedList<>(working);
            }
        }

        this.part1Answer = part1.size();
        this.part2Answer = working.size();
    }

    private LinkedList<Integer> lookSay(LinkedList<Integer> input) {
        Integer digit = input.pop();
        int count = 1;
        LinkedList<Integer> output = new LinkedList<>();

        int i = 1;
        while (input.size() > 0) {
            while (input.size() > 0 && digit == input.peekFirst()) {
                count++;
                input.pop();
            }

            output.add(count);
            output.add(digit);

            if (input.size() > 0) {
                digit = input.pop();
                count = 1;
            } else {
                count = 0;
            }
        }

        if (count > 0) {
            output.add(count);
            output.add(digit);
        }

        return output;
    }

    @Override
    public Integer getPart1Answer() {
        return part1Answer;
    }

    @Override
    public Integer getPart2Answer() {
        return part2Answer;
    }
}

File: scala-advent-of-code/src/main/scala/advent/twenty_fifteen/Day10ElvesLookElvesSay.scala
package advent.twenty_fifteen

import advent.common.DailyProblem
import advent.utilities.FileUtilities

import scala.annotation.tailrec

class Day10ElvesLookElvesSay(input: String) extends DailyProblem[Int, Int] {
  @tailrec
  private def walkString(current : Char, count : Int, remaining : List[Char], accumulator : List[(Int,Char)]) : List[(Int, Char)] = {
    if (remaining.isEmpty) {
      (count, current) :: accumulator
    } else if (remaining.head == current) {
        walkString(current, count + 1, remaining.tail, accumulator)
    } else {
      walkString(remaining.head, 1, remaining.tail, (count, current) :: accumulator)
    }
  }

  @tailrec
  private def lookSay(current : List[Char], count : Int, target : Int, accumulator: Map[Int,Int]) : Map[Int,Int] = {
    if (count > target) {
      accumulator
    } else {
      val next = walkString(current.head, 1, current.tail, List()).foldLeft(List[Char]())((a, p) => p._1.toString.charAt(0) :: p._2.toChar :: a)
      lookSay(next, count + 1, target, accumulator + (count -> current.length))
    }
  }

  private val lengths = lookSay(input.toCharArray.toList, 0, 50, Map())

  override val part1Answer: Int = lengths(40)
  override val part2Answer: Int = lengths(50)
}
